### PR TITLE
Introduces a service method

### PR DESCRIPTION
**Pull request by dmccown**
Adding a new service method where you can look up a service by origin
and name.

Signed-off-by: David McCown <dmccown@chef.io>

### CODE REVIEW

File: lib/train-habitat/httpgateway.rb [hunk 1]
```diff
@@ -15,6 +17,23 @@ def initialize(host)
       def services
         JSON.parse(Net::HTTP.get_response(uri).body)
       end
+
+      def service(origin, name)
+        service = services.select do |svc|
+          svc.dig('pkg', 'origin') == origin && svc.dig('pkg', 'name') == name
+        end
```
Comment (TrevorBramble) [inline lib/train-habitat/httpgateway.rb]: Hrm, how would you feel about...
```ruby
def service(origin, name)
  by_origin = ->(s) { s.dig('pkg', 'origin') == origin }
  by_name   = ->(s) { s.dig('pkg', 'name') == name }

  matching_services = services.select(&by_origin).select(&by_name)

  validate!(matching_services, origin, name) or service.first
end
```

File: .rubocop.yml [hunk 2]
```diff
@@ -6,3 +6,6 @@ Style/Documentation:
   Enabled: false
 Naming/FileName:
   Enabled: false
+Metrics/BlockLength:
+  Exclude:
+    - 'test/unit/httpgateway_test.rb'
```
Comment (TrevorBramble) [inline .rubocop.yml]: Should all of `./test` be excluded?

File: lib/train-habitat/illegal_state_error.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  end
+end
+
+class NoServicesFoundError < IllegalStateError
+  def initialize(origin, name)
+    super("Expected one service '#{origin}/#{name}', but found none.")
+  end
+end
+
+class MultipleservicesFoundError < IllegalStateError
```
Comment (Padgett) [inline lib/train-habitat/illegal_state_error.rb]: I'd like to see Services capitalized, ie. `MultipleServicesFoundError`

Comment (dmccown) [reply lib/train-habitat/illegal_state_error.rb]: Oops that was a typo.

File: lib/train-habitat/httpgateway.rb (hunk 1 above)
Comment (dmccown) [reply lib/train-habitat/httpgateway.rb]: I pulled out the first two lines to methods to try to keep the method at the scope. Let me know what you think.

File: lib/train-habitat/httpgateway.rb [hunk 4]
```diff
@@ -15,6 +17,34 @@ def initialize(host)
...
       end
+
+      def service(origin, name)
+        selected = services.select(&by_origin(origin))
+                           .select(&by_name(name))
+
+        validate!(selected, origin, name)
+
+        selected.first
+      end
```
Comment (TrevorBramble) [inline lib/train-habitat/httpgateway.rb]: Bugs me how procedural this is.
```ruby
def service(origin, name)
  selected = services.select(&by_origin(origin)).select(&by_name(name))

  selected.first
ensure
  validate!(selected)
end
```
You might pull out separate validation statements to clarify the logic here, and at least from my perspective even out the abstraction level.
```ruby
def service(origin, name)
  selected = services.select(&by_origin(origin)).select(&by_name(name))

  selected.first
ensure
  raise NoServicesFoundError.new(origin, name) if selected.empty?
  raise MultipleServicesFoundError.new(origin, name) if selected.size > 1
end
```

Comment (dmccown) [reply lib/train-habitat/httpgateway.rb]: When is `ensure` evaluated? If we have no services selected and try to do `selected.first` we may throw an error there.

**Pull request merged**